File: services/saas-api/src/saas_api/schemas/metrics.py
from pydantic import BaseModel, Field


class MetricsConversion(BaseModel):
    model_config = {"populate_by_name": True}

    visit_to_profile: float = Field(alias="visitToProfile")
    profile_to_report: float = Field(alias="profileToReport")
    report_to_product_click: float = Field(alias="reportToProductClick")


class DashboardMetricsResponse(BaseModel):
    model_config = {"populate_by_name": True}

    period: str
    visits: int
    onboarding_starts: int = Field(alias="onboardingStarts")
    birth_profiles_submitted: int = Field(alias="birthProfilesSubmitted")
    free_reports_requested: int = Field(alias="freeReportsRequested")
    free_reports_viewed: int = Field(alias="freeReportsViewed")
    product_clicks: int = Field(alias="productClicks")
    product_cta_clicks: int = Field(alias="productCtaClicks")
    reports_generated: int = Field(alias="reportsGenerated")
    report_failures: int = Field(alias="reportFailures")
    conversion: MetricsConversion

File: packages/api-contracts/src/metrics.ts
import { z } from "zod";

export const dashboardMetricsPeriodSchema = z.enum(["7d", "30d"]);

export const dashboardMetricsConversionSchema = z.object({
  visitToProfile: z.number(),
  profileToReport: z.number(),
  reportToProductClick: z.number(),
});

export const dashboardMetricsSchema = z.object({
  period: dashboardMetricsPeriodSchema,
  visits: z.number().int().min(0),
  onboardingStarts: z.number().int().min(0),
  birthProfilesSubmitted: z.number().int().min(0),
  freeReportsRequested: z.number().int().min(0),
  freeReportsViewed: z.number().int().min(0),
  productClicks: z.number().int().min(0),
  productCtaClicks: z.number().int().min(0),
  reportsGenerated: z.number().int().min(0),
  reportFailures: z.number().int().min(0),
  conversion: dashboardMetricsConversionSchema,
});

export type DashboardMetrics = z.infer<typeof dashboardMetricsSchema>;
export type DashboardMetricsPeriod = z.infer<typeof dashboardMetricsPeriodSchema>;
